forem/forem
# RFC#50-P5 Prepares the admin restructure to be tested

NOTE: Travis is bombing out on `ActiveRecord::NoDatabaseError: FATAL: database "Forem_test" does not exist`. Trying to figure out how to get a Travis build running, but in the meantime this PR can be reviewed.

## What type of PR is this? (check all applicable)


## Description

**It contains the following changes:**
- It fixes a bug where we was using `as:` on the scoped routes which caused the path helpers to need to prepended with the scope.
- It replaces all instances of `href` or anything manual with `link_to` so that we're able to use dynamic path helpers 
- It replaces the `render` method in controllers to reference the action within the same controller instead of referencing a path helper. 
- It uses Rails 6.1 functionality to pull out admin routes into its own file. This allows the changes to be localized and old code can be easily deleted in the future. 

**What is the goal of this PR and what was the journey**
The initial goal of this PR was to allow the old admin structure and the new one to co-exist, and for us to be able to toggle between the two with ease.

However, we encountered a pretty complex problem when trying to do this. Please read [this comment](https://github.com/forem/forem/pull/13114#issuecomment-814788111) to understand the problem we were encountering. 

**A short overview of the problem is outlined below (optional read):** 
When we toggle the feature flag it updates all the routes accordingly (i.e it adds the scope to the necessary routes when enabled), however the path helpers are not playing nicely with this change. 

When we define scoped routes, it allows us to use the same path helpers as the old routes but a different route. As an example:
- When the feature flag is enabled  the scope is added to the routes `admin/content_manager/articles` and path helper admin_articles maps to `admin/content_manager/articles`. 
- When the feature flag is disabled the behaviour is such that the route is without a scope `admin/article` and path helper admin_articles maps to `admin/articles`. 

However, it seems that with the dynamic routing (using `constraints`) it contains references to both the old routes and new scoped routes, and Rails is consuming the reference to the first path helper based on order in the routes file. The routes however are being honoured, its the path helpers that aren't.
As a result we obtaining the correct routing structure but the incorrect path helper. 

**Final decisions**
Hence, taking into consideration the following: 
- we will not be running both the new admin routes and the old admin routes together except during the testing period, after which we will delete the old routes. 
- we have exhausted many avenues to try and get the path helpers to take direction from the Feature Flag

We've decided to just use an if statement to load the correct set of routes based on boot time.  
This has the obvious problem of requiring an app reload to get the other routes after switching the flag, but since we are testing on a smaller Forem (and not DEV) this will not be an issue, and we can reload the rails app again after toggling the feature flag to get the path helpers pointing correctly.

## Related Tickets & Documents
forem/rfcs#50
and 4 other earlier PR"s

## QA Instructions, Screenshots, Recordings

- Double check that the changes I've made to any hrefs and/or renders represent the original code
-Click around the interface [by enabling the admin_restructure](https://forem-team.slack.com/archives/C7GE84DEJ/p1615240364120400) feature flag and then restarting your server. Note the description section above for details on what I've covered in this PR.

### UI accessibility concerns?

None, non-UI changes

## Added tests?

Still need to fix them.

- [x] No, and there was no new functionality added but just refactors made. 

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
      or in a [forem.dev](http://forem.dev) post
- [x] This change does not need to be communicated, and this is why not:This change does not need to be communicated, and this is why not: It will be communicated internally once ready to test.

## [optional] Are there any post deployment tasks we need to perform?
Yes the feature flag on a smaller Forem needs to be toggled and the rails app reloaded. 

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/xTk9ZKVGEptejVquSk/giphy.gif)
